### ISSUE TITLE
mysqli: error on invalid escaping

### ISSUE THREAD

**Issue by staabm** [staabm/phpstan-dba#562]
wrong:
```
$query = sprintf('SELECT CountryCode FROM City WHERE name="%s"', // double quotes in query
    mysqli_real_escape_string($mysqli, $city));

$query = sprintf('SELECT CountryCode FROM City WHERE name="%s"', // double quotes in query
    $mysqli->real_escape_string($city));
```

correct:
```
$query = sprintf("SELECT CountryCode FROM City WHERE name='%s'",
    $mysqli->real_escape_string($city));
```

**Comment by staabm** [comment]
@craigfrancis do you agree that these are cases we should report errors for, as they hide errors in wrong escaping?

do other php apis have the same problem? 
I checked PDO, but `PDO::quote` already wrapps the returned value in single-quotes, so at least the mentioned examples shouldn't be a problem for PDO.

**Comment by staabm** [comment]
I think I have seen similar code like this in my life somewhere:

wrong (sql injection vulnerable):
```php
$query = sprintf('SELECT CountryCode FROM City WHERE name="%s"', // double quotes in query
    addslashes($mysqli, $city));

$query = sprintf('SELECT CountryCode FROM City WHERE name="%s"', // double quotes in query
    addcslashes($city));

$query = sprintf('SELECT CountryCode FROM City WHERE name=%s', // no quotes in query
    mysqli_real_escape_string($mysqli, $city));

$query = sprintf('SELECT CountryCode FROM City WHERE name=%s', // no quotes in query
    $mysqli->real_escape_string($city));
```

**Comment by craigfrancis** [comment]
Yep, missing quotes is a common problem... a code base with many thousands of values being escaped, and all it takes is 1 mistake.

The `mysqli_real_escape_string()` function is dangerous, developers learn about escaping, and they use it, and think they are now safe... it's one of the main reasons why I keep pushing developers to use database abstractions, or parameterised queries (both still need to use the [`literal-string` type](https://phpstan.org/writing-php-code/phpdoc-types#other-advanced-string-types) to check mistakes haven't been made).

It's even more fun when developers try escaping field names, e.g. ```$sql .= ' ORDER BY `' . $db->real_escape_string($field) . '`'``` (and yes, I've seen it without the backtick quotes as well).

As to similar API's...

- PDO, yes, you're correct, `PDO::quote` is much safer because it adds the quotes itself, but I will note the [manual](https://www.php.net/manual/en/pdo.quote.php) says "you are strongly recommended to use PDO::prepare() to prepare SQL statements with bound parameters".
- PostgreSQL, the `pg_escape_string()` has the missing quotes issue, and you also need to use single quotes as well (double quotes don't get escaped, [source](https://git.postgresql.org/gitweb/?p=postgresql.git;a=blob;f=src/interfaces/libpq/fe-exec.c;h=ec62550e3858135a48be0489df51883c0f9b25ca;hb=HEAD#l4015)), that's why you should at least use `pg_escape_literal()`... or, please, just use parameterised queries :-)
- MSSQL, the `mssql_*` functions were removed in PHP 7, but I seem to remember it didn't have a string escaping function, I believe they were trying to get everyone to use parameterised queries. Unfortunately I've seen developers make their own, from `str_replace('"', '""', str_replace("'", "''", $string))` to using a [hex bytestring](https://www.appsloveworld.com/sql-server/100/2/how-to-escape-strings-in-sql-server-using-php).
- SQLSRV, the replacement for MSSQL, that also pushes developers to use prepared queries.
- HTML encoding has a similar issue `"<img src=" . htmlentities($url) . " alt='' />"`, where $url could be set to `'/ onerror=alert(1)'` (and yes, it does not consider `javascript:` urls in `<a href="">`).
- Fortunately `escapeshellarg()` does add single quotes.
- And when I find `addslashes()` being used incorrectly, I should put a part of those invoice payments into a holiday fund, and go on a lot more holidays in it's (dis)honor.

And when you say you have seen similar code before, we kinda discussed this last year, ref [issue 316](https://github.com/staabm/phpstan-dba/issues/316).

In your examples, a `name=%s` is most likely going to be picked up fairly quickly (usually an SQL parse error), I tend to find this mistake takes the form of `id=%s`, because a number is normally being passed to it, and that works ™.

Extra bonus point for parameterised queries... the developer defined SQL is sent to the database first, the query execution plan is created from it (sometimes that can be cached)... then the user values are accepted/used; this means the user values cannot have any affect on the query execution plan, so even if there was an implementation issue, it's far too late to change it... in contrast, (escaped) user values are mixed in with the SQL, and separating them again is tricky.

**Comment by staabm** [comment]
thanks. I agree that prepared statements is a must have. still there is a lot of vulnerable code out there, where I want to point out security issues.

I will add a "phpstan-tip" to the rule that escaping should be prevented. you had a website somewhere, where all the problems and solutions are described in detail, right?

**Comment by craigfrancis** [comment]
That works.

As to the website, yep, although it's more of a quick summary (so people are more likely to read): https://eiv.dev/

It does link to pages on my GitHub repo with some typical examples:

https://github.com/craigfrancis/php-is-literal-rfc/blob/main/justification/escaping.php
https://github.com/craigfrancis/php-is-literal-rfc/blob/main/justification/mistakes.php

And yes, there is a lot of vulnerable code out there, and why I'm so glad you're tackling it from another angle... hopefully, with enough people working on this problem, we can stop boring injection vulnerabilities. and I can move on to more interesting problems :-)
